### PR TITLE
Per weapon auto convergence

### LINKED ISSUE #4947: Feature Request: Per-Weapon convergences.
Heya

Seeing @MjnMixael implement per-weapon auto-aim had me thinking: What about per-weapon convergences? Would love to be able to, for example, have each weapon convergence at its maximum range.

### PR DESCRIPTION
**Pull request by MjnMixael**
This adds per-weapon convergence features, notably allowing per-weapon autoaim + convergence. I moved the autoaim flags to Object:: from Ship:: so they could be reused. Seemed the most sensible place.. but could go elsewhere if there's a better spot.

While technically a new feature, this allows BtA to fix an issue with one of our weapons.. so really hoping to get this in for 24.2.

Fixes #4947

### CODE REVIEW

File: code/ship/ship.cpp [hunk 1]
```diff
@@ -12458,9 +12458,9 @@ int ship_fire_primary(object * obj, int force, bool rollback_shot)
...
-		has_autoaim = ((has_converging_autoaim || (sip->aiming_flags[Ship::Aiming_Flags::Autoaim]) || (winfo_p->autoaim_fov > 0.0f) || (sip->bank_autoaim_fov[bank_to_fire] > 0.0f)) && aip->target_objnum != -1);
-		needs_target_pos = ((has_autoaim || (sip->aiming_flags[Ship::Aiming_Flags::Auto_convergence])) && aip->target_objnum != -1);
+		has_converging_autoaim = ((sip->aiming_flags[Object::Aiming_Flags::Autoaim_convergence] || winfo_p->aiming_flags[Object::Aiming_Flags::Autoaim_convergence] || (The_mission.ai_profile->player_autoaim_fov[Game_skill_level] > 0.0f && !( Game_mode & GM_MULTIPLAYER ))) && aip->target_objnum != -1);
+		has_autoaim = ((has_converging_autoaim || (sip->aiming_flags[Object::Aiming_Flags::Autoaim]) || winfo_p->aiming_flags[Object::Aiming_Flags::Autoaim] || (sip->bank_autoaim_fov[bank_to_fire] > 0.0f)) && aip->target_objnum != -1);
+		needs_target_pos = ((has_autoaim || has_converging_autoaim) && aip->target_objnum != -1); // This change needs some scrutiny but it seems right?
```
Comment (wookieejedi) [inline code/ship/ship.cpp:12463]: Good question. In current master this part of the check is `(sip->aiming_flags[Ship::Aiming_Flags::Auto_convergence])` and in the code here it is replaced with `has_converging_autoaim`.

`(sip->aiming_flags[Ship::Aiming_Flags::Auto_convergence])` is convergence with no autoaim on, and thus not the same as `has_converging_autoaim`. I would say using `(sip->aiming_flags[Object::Aiming_Flags::Auto_convergence]) || winfo_p->aiming_flags[Object::Aiming_Flags::Auto_convergence])` would fit as a more proper replacement. (Unless I'm missing something). What do you think?

Comment (MjnMixael) [reply code/ship/ship.cpp:12463]: Ah yes. Good catch. I conflated auto_convergence with autoaim_convergence!

Comment (wookieejedi) [reply code/ship/ship.cpp:12463]: Happy to help!

Comment (MjnMixael) [reply code/ship/ship.cpp:12463]: I'm going to break down these bools for readability, I think. There's just a lot of checks here. The complier should optimize these out anyway.

Comment (MjnMixael) [reply code/ship/ship.cpp]: Ok, give that a look. I think that's a lot more readable

File: code/ship/ship.cpp [hunk 2]
```diff
@@ -12458,9 +12463,17 @@ int ship_fire_primary(object * obj, int force, bool rollback_shot)
...
+		
+		// This logic is getting a little unweildy.. so let's break it down slightly for readability.
+		// First check our ship/weapon flags
+		const bool autoaim_flagged = ((sip->aiming_flags[Object::Aiming_Flags::Autoaim]) || (winfo_p->aiming_flags[Object::Aiming_Flags::Autoaim]));
+		const bool autoaim_convergence_flagged = ((sip->aiming_flags[Object::Aiming_Flags::Autoaim_convergence]) || (winfo_p->aiming_flags[Object::Aiming_Flags::Autoaim_convergence]));
+		const bool auto_convergence_flagged = ((sip->aiming_flags[Object::Aiming_Flags::Auto_convergence]) || (winfo_p->aiming_flags[Object::Aiming_Flags::Auto_convergence]));
+
+		// Now add in other checks like difficulty and game mode as appropriate. Also make sure we have a target and an FOV if necessary.
+		has_converging_autoaim = ((autoaim_convergence_flagged || (The_mission.ai_profile->player_autoaim_fov[Game_skill_level] > 0.0f && !( Game_mode & GM_MULTIPLAYER ))) && aip->target_objnum != -1);
+		has_autoaim = ((autoaim_flagged || has_converging_autoaim || (sip->bank_autoaim_fov[bank_to_fire] > 0.0f)) && aip->target_objnum != -1);
```
Comment (wookieejedi) [inline code/ship/ship.cpp:12475]: Looks good, just want to double check on one more:
Original: 
`has_autoaim = ((has_converging_autoaim || (sip->aiming_flags[Ship::Aiming_Flags::Autoaim]) || (winfo_p->autoaim_fov > 0.0f) || (sip->bank_autoaim_fov[bank_to_fire] > 0.0f)) && aip->target_objnum != -1);`

New:
`has_autoaim = ((autoaim_flagged || has_converging_autoaim || (sip->bank_autoaim_fov[bank_to_fire] > 0.0f)) && aip->target_objnum != -1);`

Looks like the new check is missing `(winfo_p->autoaim_fov > 0.0f)`?


Other very nitpicky thing, should `an FOV` by `a FOV` in the comments?

Comment (MjnMixael) [reply code/ship/ship.cpp:12475]: Yes that one's correct. Previously we checked that the weapon autoaim_fov > 0 to determine if we had autoaim on the weapon. Now that the weapon has the aiming flags we can check the flag directly. The autoaim_fov is used later to determine what fov to actually aim with.

Comment (MjnMixael) [reply code/ship/ship.cpp:12475]: "an FOV" when spoken aloud is how I'd say it... typed it's probably "a FOV". Either way probably not worth a whole commit just to change that.

Comment (wookieejedi) [reply code/ship/ship.cpp:12475]: Sounds all good to me then!

File: code/ship/ship.cpp (hunk 1 above)
Comment (wookieejedi) [reply code/ship/ship.cpp]: Thanks, that's definitely a good change.

**Review by wookieejedi — APPROVED** [review]
Previous comments addressed and my tests show things working as expected.

**Review by BMagnu — APPROVED** [review]

**Pull request merged**